### ISSUE TITLE
Re-rendering after scroll does not work on Epos 2 (

### ISSUE THREAD

**Issue by tolinobugreport** [koreader/android-luajit-launcher#392]

**Comment by Frenzie** [comment]
My guess is that either means the device has an issue with one of these waveform modes or it drops one of the calls.

https://github.com/koreader/android-luajit-launcher/blob/ee4cca8b81b4bd9cf3a299edb920d41cbfcd5a5a/app/src/main/java/org/koreader/launcher/device/epd/TolinoEPDController.kt

**Comment by tolinobugreport** [comment]
This does not happen always, but on some occasions every N-th scroll is garbled, on other occasions, only the N-th scroll is OK.
Opening a menu, thereby forcing a re-rendering after it is closed fixes the issue (workaround)

So, after stopping the scroll, the text would have to be forcefully re-rendered in the same way like opening a menu.

**Comment by poire-z** [comment]
You can also test with a small diagonal swipe, that does a full screen refresh, if it's enough to get the thing to display correctly (which may help pinpointing at which level is the issue).

**Comment by tolinobugreport** [comment]
Photo: [](url)
![photo_2022-11-24_13-12-48](https://user-images.githubusercontent.com/118958359/203782197-30d609e9-3a7d-4a42-829c-8f716445b514.jpg)


Download video: https://workupload.com/file/KSRJss56CTS

**Comment by Frenzie** [comment]
@tolinobugreport 
> So, after stopping the scroll, the text would have to be forcefully re-rendered in the same way like opening a menu.

Yes, that is exactly what happens. ;-) What I was trying to convey is that the most likely explanation is that the device is buggy and drops the call. The bigger question is how easy it is to work around.

**Comment by tolinobugreport** [comment]
> You can also test with a small diagonal swipe, that does a full screen refresh, if it's enough to get the thing to display correctly (which may help pinpointing at which level is the issue).

This fixes it - i.e. ungarbles the text in EPUB or "smoothens" it in PDFs.     This shoudl happen automatically after stopping the scroll.

**Comment by Frenzie** [comment]
Then the device drops the call, even though otoh we do something like a 250 ms wait to prevent that kind of thing from happening on bad devices. (Perhaps Tolino isn't marked as a bad device.)

**Comment by pazos** [comment]
> Then the device drops the call, even though otoh we do something like a 250 ms wait to prevent that kind of thing from happening on bad devices. (Perhaps Tolino isn't marked as a bad device.)

Delay is added on devices where we race the system. It is only applied to fullscreen updates. So we let the system to end the shitty update and perform a full update after it. It doesn't apply the tolino.

> Yes, that is exactly what happens. ;-) What I was trying to convey is that the most likely explanation is that the device is buggy and drops the call. The bigger question is how easy it is to work around. @pazos will know better.

It should be workaround-able. But requires somebody with the device and a bit of knowledge to tinker with the [TolinoEPDController](https://github.com/koreader/android-luajit-launcher/blob/ee4cca8b81b4bd9cf3a299edb920d41cbfcd5a5a/app/src/main/java/org/koreader/launcher/device/epd/TolinoEPDController.kt)

If the call that is droped is WaveformFull then add some delay  and see what happens.
If the call that is droped is WaveformFullUi it could mean that `EINK_WAVEFORM_MODE_GLR16` doesn't apply to that particular device and can be replaced with `EINK_WAVEFORM_MODE_GC16`

I won't attempt to solve it. Help wanted from tolino users.

**Comment by tolinobugreport** [comment]
I could reproduce the issue also on the older Tolino Page 1:

![photo_2022-11-24_15-23-16](https://user-images.githubusercontent.com/118958359/203807384-b1abe7f2-622c-44d1-97e1-c78b295be86a.jpg)

**Comment by poire-z** [comment]
There are different Scrolling option in `Tap & gestures > Scrolling>`. Most will need many screen refreshes to give the expected visual effect. "On release scrolling" will just issue a refresh at the end, so many less.
Again, you might want to check how it does - and use that in the meantime.

**Comment by tolinobugreport** [comment]
After upgrading to the latest 2022.11, it seems that the problem is gone...    Corruption still happens **during** scrolling which is acceptable/expected, but then after stopping the scroll, the page refreshes correctly.

So far I have not been able to reproduce the issue on the Tolino Epos 2 with 2022.11

More testing is required tomorrow on the Tolino Page 1.
